### PR TITLE
Fix topology serialization / reinitialization

### PR DESCRIPTION
**Pull request by mattwthompson**
**In the refactor branch only** I found I couldn't re-initialize a `Topology` if it already had box vectors, and after some digging I believe the cause is me not considering this case when I re-wrote the serialization methods. There was also some bad logic that slipped in, mostly because there wasn't yet a test for this behavior.

```
In [1]: from openff.toolkit.topology import Topology

In [2]: from openff.units import unit

In [3]: foo = Topology()

In [4]: foo.box_vectors = [1, 2, 3] * unit.nanometer

In [5]: Topology(foo)
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
~/miniconda3/envs/openff-dev/lib/python3.7/site-packages/pint/quantity.py in __getattr__(self, item)
   1889         try:
-> 1890             return getattr(self._magnitude, item)
   1891         except AttributeError:

AttributeError: 'float' object has no attribute '_value'

During handling of the above exception, another exception occurred:

AttributeError                            Traceback (most recent call last)
<ipython-input-5-7a14f5f6a055> in <module>
----> 1 Topology(foo)

~/software/openforcefield/openff/toolkit/topology/topology.py in __init__(self, other)
    424         # TODO: Try to construct Topology copy from `other` if specified
    425         if isinstance(other, Topology):
--> 426             self.copy_initializer(other)
    427         elif isinstance(other, FrozenMolecule):
    428             self.from_molecules([other])

~/software/openforcefield/openff/toolkit/topology/topology.py in copy_initializer(self, other)
   1277     def copy_initializer(self, other):
   1278         other_dict = copy.deepcopy(other.to_dict())
-> 1279         self._initialize_from_dict(other_dict)
   1280
   1281     def to_dict(self):

~/software/openforcefield/openff/toolkit/topology/topology.py in _initialize_from_dict(self, topology_dict)
   1325             self._box_vectors = None
   1326         else:
-> 1327             self.box_vectors = string_to_quantity(topology_dict["box_vectors"])
   1328         for molecule_dict in topology_dict["molecules"]:
   1329             new_mol = Molecule.from_dict(molecule_dict)

~/software/openforcefield/openff/toolkit/topology/topology.py in box_vectors(self, box_vectors)
    575                     f"Box vectors must be shape (3, 3). Found shape {box_vectors.shape}"
    576                 )
--> 577         elif isinstance(box_vectors._value, list):
    578             if len(box_vectors) == 3:
    579                 box_vectors._value *= np.eye(3)

~/miniconda3/envs/openff-dev/lib/python3.7/site-packages/pint/quantity.py in __getattr__(self, item)
   1892             raise AttributeError(
   1893                 "Neither Quantity object nor its magnitude ({}) "
-> 1894                 "has attribute '{}'".format(self._magnitude, item)
   1895             )
   1896

AttributeError: Neither Quantity object nor its magnitude (0.0) has attribute '_value'

In [6]: foo.box_vectors = None

In [7]: Topology(foo)
Out[7]: <openff.toolkit.topology.topology.Topology at 0x7f88343099d0>
```

This operation is going to be common in Interchange; one of the first things that happens in `Interchange.from_smirnoff` is deepcopying the input topology. This should prevent bad behavior like reported in https://github.com/openforcefield/openff-interchange/issues/141 and also enable the "pre-" and "post-parameterization" dual topology design discussed in a few places (#748 ?)
- [x] Reproduction
- [x] Add [tests](https://github.com/openforcefield/openff-toolkit/tree/master/openff/toolkit/tests)
- [ ] Update docstrings/[documentation](https://github.com/openforcefield/openff-toolkit/tree/master/docs), if applicable
- [x] [Lint](https://open-forcefield-toolkit.readthedocs.io/en/latest/developing.html#style-guide) codebase
- [ ] Update [changelog](https://github.com/openforcefield/openff-toolkit/blob/master/docs/releasehistory.rst)

### CODE REVIEW

**Comment by j-wags** [comment]
Nice. Looks like there's an error with element parsing now - had we decided to just compare element symbols in these cases? Either way, I agree with the major goal of this PR. Let me know when it's ready for review!

**Comment by mattwthompson** [comment]
This has been a fun bit of spelunking, actually. I worked backwards for a while with the objective of getting tests to pass before I noticed that the tests use OpenMM elements as inputs but the `Atom` class now bears elements from `mendeleev`. I was inadvertently doing some comparisons against their implementation, slowly working around strange edge cases in the heavier elements, i.e. one such case

```python3
In [1]: from mendeleev import element

In [2]: from openmm.app import element as openmm_element

In [3]: element(openmm_element.Element.getByMass(264).atomic_number).mass
Out[3]: 274.0
```

I decided it's not worth our time to worry about these differences; in my poking around I did not see anything alarming for elements smaller than 100 atomic number. The tests now do not spend time investigating these comparisons, `TestAtom::test_atom_properties` is only meant to be a test of the constructor and getters.

I updated some other tests but have not run the test suite locally to see if anything new breaks. It's worth another pass through other unit tests to see if similar changes need to be made elsewhere. I'll ping you when it's ready!

**Comment by mattwthompson** [comment]
With some haphazard grepping, I couldn't find anywhere else that uses `openmm.app.element`, so I think this is ready to go.

**Review by j-wags — APPROVED** [review]
Looks great. Thanks for the fix, @mattwthompson. One comment on a bug that wasn't introduced here, but it may be good to add a note/TODO to the code for future development

File: openff/toolkit/topology/topology.py [hunk 1]
```diff
@@ -574,9 +574,13 @@ def box_vectors(self, box_vectors):
                 raise InvalidBoxVectorsError(
                     f"Box vectors must be shape (3, 3). Found shape {box_vectors.shape}"
                 )
-        elif isinstance(box_vectors._value, list):
+        elif isinstance(box_vectors, list):
```
[the remark this answers is not in the document]
Comment (j-wags) [reply openff/toolkit/topology/topology.py:577]:
```suggestion
        # TODO: Add logic to check whether this is a 3x3 list of lists and tell use user to provide a 3x3 numpy array instead
        elif isinstance(box_vectors, list):
```

[the remark this answers is not in the document]
Comment (mattwthompson) [reply openff/toolkit/topology/topology.py:578]: I think it's reasonable a user might try to do that, so I'll add a test for it. It'll probably choke with this code, so I'll also look at refactoring the logic around lists.

[the remark this answers is not in the document]
Comment (mattwthompson) [reply openff/toolkit/topology/topology.py]: So ... this clause wasn't actually hit, since `pint.Quantity` converts lists to array-like, resulting it bearing the `shape` attribute. Check out L577 [here](https://app.codecov.io/gh/openforcefield/openff-toolkit/blob/0dbf79be8a8b0768694a88f199d1f815dbe2274d/openff/toolkit/topology/topology.py).

I've added a couple of tests cases to safeguard against other cases, which are now all passing (locally).
